### PR TITLE
add capture group name in comment

### LINKED ISSUE #62715: Regex source generator should include capture group name in comment
Given an expression like:
```C#
"^(?<LINE>[0-9]*)-(?<ENDLINE>[0-9]*)$"
```
the regex source generator will output comments for the groups like:
```C#
// 1st capturing group
```
We should include the name if one exists, e.g.
```C#
// 1st capturing group ("LINE")
```

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #62715

### CODE REVIEW

**Comment by GSPP** [comment]
Can capture group names contain newline characters? That could break the code that's emitted.

**Comment by pedrobsaila** [comment]
> Can capture group names contain newline characters? That could break the code that's emitted.

Newline characters are not valid inside group name, if you try it inside a console app, you will get a RegexParseException with message : "Invalid pattern '123 (?<group
name>ABC) 123' at offset 12. Invalid group name: Group names must begin with a word character."

File: src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs [hunk 1]
```diff
@@ -556,6 +556,46 @@ partial class C
...
         }
 
+        [Fact]
+        public async Task WithGroups_ShouldDisplayGroupName()
+        {
+            string code = @"
+                using System.Text.RegularExpressions;
+                partial class C1
+                {
+                    [RegexGenerator(""^(?<LINE>[0-9]*)-(?<ENDLINE>[0-9]*)$"")]
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:566]: should this include an unnamed and a numbered capture as well

File: src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs [hunk 2]
```diff
@@ -556,6 +556,46 @@ partial class C
...
+               .WithMetadataReferences(s_refs)
+               .WithCompilationOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
+               .WithNullableContextOptions(NullableContextOptions.Enable))
+               .WithParseOptions(new CSharpParseOptions(langVersion))
+               .AddDocument("RegexGenerator.g.cs", SourceText.From(code, Encoding.UTF8)).Project;
+
+            Assert.True(proj.Solution.Workspace.TryApplyChanges(proj.Solution));
+
+            Compilation? comp = await proj!.GetCompilationAsync(CancellationToken.None).ConfigureAwait(false);
+            Debug.Assert(comp is not null);
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:583]: It's unusual to use Debug.Assert in tests. We generally want them to fail in retail as well. Consider Assert.NotNull instead.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 3]
```diff
@@ -1666,7 +1666,8 @@ void EmitNode(RegexNode node, RegexNode? subsequent = null, bool emitLengthCheck
 
                 // Put the node's code into its own scope. If the node contains labels that may need to
                 // be visible outside of its scope, the scope is still emitted for clarity but is commented out.
-                using (EmitScope(writer, DescribeNode(node), faux: PossiblyBacktracks(node) && !node.IsAtomicByParent()))
+                string capName = ExtractCapName(node, rm.Code.Tree.CapNames);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:1669]: We should only do this inside of DescribeNode in the case where the node is actually a capture.  I suggest adding a DescribeCapture helper function that DescribeNode uses; it is then responsible for formatting the capture description, including looking up any name that's needed for the rendering.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 4]
```diff
@@ -3372,7 +3373,7 @@ node.Type switch
                 RegexNode.Bol => "Match if at the beginning of a line.",
                 RegexNode.Boundary => $"Match if at a word boundary.",
                 RegexNode.Capture when node.N != -1 => $"{DescribeNonNegative(node.M)} capturing group. Uncaptures the {DescribeNonNegative(node.N)} capturing group.",
-                RegexNode.Capture when node.N == -1 => $"{DescribeNonNegative(node.M)} capturing group.",
+                RegexNode.Capture when node.N == -1 => $"{DescribeNonNegative(node.M)} capturing group" + (string.IsNullOrEmpty(capName) ? "." : $" (\"{capName}\")."),
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3376]: This should use the `Literal(...)` helper to output the `capName`.  That not only handles adding the quotes, it ensures characters in the string are rendered appropriately for C# source.

We also need to do a similar rendering for the case when node.N != -1, both for the capture group and the uncapture group.

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 5]
```diff
@@ -3436,11 +3439,27 @@ private static void DescribeExpression(TextWriter writer, RegexNode node, string
...
+        /// <param name="capNames">cap name table</param>
+        /// <returns></returns>
+        private static string ExtractCapName(RegexNode node, Hashtable capNames)
+        {
+            if (node.Type != RegexNode.Capture || capNames == null || capNames.Count == 0)
+            {
+                return string.Empty;
+            }
+
+            return capNames.Keys.OfType<string>().FirstOrDefault(i => (int)capNames[i] == node.M);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3460]: We shouldn't use LINQ here.  This should use the same lookup logic that's used in https://github.com/dotnet/runtime/blob/55fecfbc201dc789d71c0903e04f2a779212c671/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs#L341-L355
That logic could even be refactored out into a helper that both Regex and this then call, e.g. putting it into an internal static on RegexParser or somewhere similar.

File: src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs [hunk 6]
```diff
@@ -556,6 +556,46 @@ partial class C
...
+            GeneratorDriverRunResult generatorResults = gd.GetRunResult();
+            Assert.NotNull(generatorResults);
+            GeneratorRunResult generatorRunResult = Assert.Single(generatorResults.Results);
+            GeneratedSourceResult generatedSourceResult = Assert.Single(generatorRunResult.GeneratedSources);
+            string sourceText = generatedSourceResult.SourceText.ToString();
+            Assert.Contains("// 1st capturing group (\"LINE\").", sourceText);
+            Assert.Contains("// 2nd capturing group (\"ENDLINE\").", sourceText);
+            Assert.Contains("// \u25CB 1st capturing group (\"LINE\").", sourceText);
+            Assert.Contains("// \u25CB 2nd capturing group (\"ENDLINE\").", sourceText);
+        }
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:597]: I applaud your adding a test.  However, I don't want to duplicate all of this workspace/generator/compilation logic: if the existing helper is insufficient to perform the desired validation, we should refactor it to be suitable.  We also don't currently have tests that validate any other aspects of code generation.  I believe we have an open issue to do that, but regardless, @joperezr has it on his plate to add a bunch of tests that validate the code generated for a given set of input expressions is exactly what we expect, and that will naturally handle this case as well as long as the input expressions use named capture groups (part of his work will be ensuring we have good coverage of all features in the tested expressions). Thus, while normally I encourage adding unit tests, I'm ok with not having one in this PR.

**Comment by danmoseley** [comment]
> Newline characters are not valid inside group name. If you try it inside a console app, you will get a RegexParseException with message : "Invalid pattern '[regex]' Invalid group name: Group names must begin with a word character."

That message is wrong, it should say "Invalid pattern '[regex]' Invalid group name: Group names must only contain word characters.".

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs (hunk 5 above)
Comment (pedrobsaila) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3460]: I'm ok with dropping LINQ but I don't think this logic is the one needed :
* Here is an example of what caps look like with Regex.IsMatch("10-563", "^(?<LINE>[0-9]*)-(?<ENDLINE>[0-9]*)$") when calling GroupNameFromNumber
image
* Here is what caps look like in my case
image

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:3460]: This:
```C#
using System.Text.RegularExpressions;

Regex r = new Regex("^(?<LINE>[0-9]*)-(?<ENDLINE>[0-9]*)$");
Console.WriteLine(r.GroupNameFromNumber(0));
Console.WriteLine(r.GroupNameFromNumber(1));
Console.WriteLine(r.GroupNameFromNumber(2));
```
outputs:
```
0
LINE
ENDLINE
```

**Comment by stephentoub** [comment]
Thanks, @pedrobsaila.  I rebased your branch and pushed a commit to fix up a few things.  Should be good to go once CI passes.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**